### ISSUE TITLE
feat(Φ5): trios-training — IGLA-GF16 training pipeline for Parameter Golf submission

### ISSUE THREAD

**Issue by gHashTag** [gHashTag/trios#22]
## 🎯 Цель

Реализовать полный тренировочный пайплайн **IGLA-GF16** для сабмита в [OpenAI Parameter Golf](https://github.com/openai/parameter-golf) до **30 апреля 2026**.

Базируется на готовых модулях Φ1–Φ4 (precision-router, trios-golden-float, trios-ternary, trios-hybrid).

---

## 📐 Архитектурные константы (IGLA-GF16)

```rust
const PHI: f64 = 1.6180339887498948482;
const ALPHA_PHI: f64 = 0.11803398874989483; // φ⁻³/2 = α_s(mZ) PDG2024

// Fibonacci architecture
const D_MODEL:  usize = 144;   // Fib #12
const N_HEADS:  usize = 8;     // Fib #6
const D_FFN:    usize = 233;   // Fib #13 ≈ 144×φ
const N_LAYERS: usize = 7;     // 7 layers → fits 16MB
const VOCAB:    usize = 32000; // LLaMA tokenizer

// Size check: 32000×144×2B + 7×(4×144²+2×144×233)×2B ≈ 9.5MB ✅
```

---

## 📦 Структура крейта: `crates/trios-training/`

```
crates/trios-training/
├── Cargo.toml
└── src/
    ├── lib.rs
    ├── model.rs          # IGLA-GF16 transformer (7 layers, Fib dims)
    ├── phi_schedule.rs   # φ-LR: LR(t) = α_φ · φ^(−t/τ), τ=T/(φ·27)
    ├── trinity_init.rs   # 4-sector weight init (gauge/higgs/lepton/cosmology)
    ├── ca_mask.rs        # Fibonacci sparse attention mask
    ├── data.rs           # FineWeb tokenized batch loader
    ├── train.rs          # training loop + BPB metric
    └── eval.rs           # Parameter Golf eval format output
```

---

### Stage 1: Scaffold (Cargo.toml + lib.rs)
- [ ] Добавить `trios-training` в workspace `Cargo.toml`
- [ ] Зависимости: `trios-golden-float`, `trios-ternary`, `trios-hybrid`, `candle-core` или pure-Rust матмул
- [ ] `lib.rs` с pub re-exports всех модулей
- [ ] **НЕТ .sh файлов** — только Rust/TypeScript

### Stage 2: Model (`model.rs`)
- [ ] `IglaGf16Config` struct с Fibonacci константами
- [ ] `IglaGf16Model` — 7-слойный трансформер
- [ ] Embedding (tied weights) → GF16 (через `trios-golden-float`)
- [ ] Attention (8 heads, d_head=18) → GF16 + CA φ-mask
- [ ] FFN (144→233→144) → Ternary (через `trios-ternary`)
- [ ] Output projection → GF16
- [ ] `model.size_bytes() -> usize` — должно быть ≤ 16_777_216

### Stage 3: φ-LR Schedule (`phi_schedule.rs`)
```rust
pub fn phi_lr(step: usize, max_steps: usize) -> f64 {
    let tau = max_steps as f64 / (PHI * 27.0);
    ALPHA_PHI * PHI.powf(-(step as f64 / tau))
}
pub fn phi_lr_warmup(step: usize, warmup: usize, max_steps: usize) -> f64 {
    if step < warmup {
        ALPHA_PHI * (step as f64 / warmup as f64)
    } else {
        phi_lr(step - warmup, max_steps - warmup)
    }
}
```
- [ ] Тест: `phi_lr(0, 10000) == ALPHA_PHI` (Δ < 1e-10)
- [ ] Тест: warmup за Fib(7)=21 шагов

### Stage 4: Trinity Weight Init (`trinity_init.rs`)
```rust
pub enum PhysicsSector { Gauge, Higgs, Lepton, Cosmology }

pub fn trinity_std(sector: PhysicsSector) -> f64 {
    match sector {
        Gauge     => ALPHA_PHI,                    // 0.11803 — attn QKV
        Higgs     => ALPHA_PHI * (PHI - 1.0),      // 0.07295 — attn proj
        Lepton    => ALPHA_PHI * (PHI - 1.0).powi(2), // 0.04508 — ffn
        Cosmology => ALPHA_PHI * (PHI - 1.0).powi(3), // 0.02786 — embed
    }
}
```
- [ ] Тест: все 4 std совпадают с расчётными значениями (Δ < 1e-8)
- [ ] Тест: Trinity Identity `PHI² + PHI⁻² = 3.0` (Δ < 1e-12)

### Stage 5: CA φ-Mask (`ca_mask.rs`)
- [ ] Fibonacci distances: `{1,2,3,5,8,13,21,34,55,89,144}`
- [ ] `build_phi_mask(seq_len: usize) -> Vec<Vec<bool>>`
- [ ] Sparsity check: для T=512 → 2.15% (11/512)
- [ ] Тест: токен 0 видит только себя, токен 200 видит ровно 11 предыдущих

### Stage 6: Data Loader (`data.rs`)
- [ ] Читать pre-tokenized FineWeb из `data/fineweb/` (бинарные .bin файлы)
- [ ] `DataBatch { input_ids: Vec<u32>, seq_len: usize }`
- [ ] Shuffle + chunk по 512 токенов
- [ ] Тест: загрузка 1 батча из синтетических данных

### Stage 7: Training Loop (`train.rs`)
```rust
pub struct TrainConfig {
    pub max_steps: usize,     // 10 мин @ 8×H100 ≈ ~50_000 steps
    pub batch_size: usize,    // 32
    pub seq_len: usize,       // 512
    pub warmup_steps: usize,  // 21 (Fib #7)
    pub grad_clip: f64,       // φ⁻¹ = 0.618
}
```
- [ ] Forward pass → cross-entropy loss
- [ ] BPB metric: `loss / ln(2)` (bits-per-byte)
- [ ] Log каждые Fib(8)=34 шага: `step, lr, loss, bpb`
- [ ] Checkpoint: сохранить веса в GF16 формате ≤ 16MB

### Stage 8: Eval (`eval.rs`)
- [ ] Совместимость с `openai/parameter-golf` eval format
- [ ] `generate_submission_json()` — метаданные модели
- [ ] Финальный BPB на validation split FineWeb

---

## 🏆 Acceptance Criteria

- [ ] `cargo test -p trios-training` — все тесты зелёные
- [ ] `model.size_bytes() <= 16_777_216` (16MB)
- [ ] `phi_lr(0) == ALPHA_PHI` (Δ < 1e-10)
- [ ] Trinity Identity: `PHI² + PHI⁻² = 3.0` (Δ < 1e-12)
- [ ] CA mask sparsity = 2.15% для T=512
- [ ] BPB лог выводится в stdout каждые 34 шага
- [ ] Checkpoint файл ≤ 16MB в GF16
- [ ] **НЕТ .sh файлов**
- [ ] Все изменения через linked issue (этот Issue #)

---

## 🔗 Зависимости

- Φ1: `trios` precision-router — [8458b60](https://github.com/gHashTag/trios/commit/8458b601e8c5f696c55ed643fa1970104251cf46)
- Φ2: `trios-golden-float` — [27777bb](https://github.com/gHashTag/trios/commit/27777bbce3b63d60954b7514c8add806e4716a95)
- Φ3: `trios-ternary` — [5f4de59](https://github.com/gHashTag/trios/commit/5f4de5911e17c0f3fcff353c3bc51fab4876f8fc)
- Φ4: `trios-hybrid` — [40125c5](https://github.com/gHashTag/trios/commit/40125c5d44bfa157829da8333ec69fc8f8a5e6b8)
- GF16 whitepaper: [zig-golden-float](https://github.com/gHashTag/zig-golden-float/blob/main/docs/whitepaper.md)
- Trinity paper: §11 IGLA-GF16 Architecture

---

## ⏰ Дедлайн

**30 апреля 2026** — сабмит в OpenAI Parameter Golf.
Осталось **11 дней**.

Priority: 🔴 CRITICAL PATH

**Comment by gHashTag** [comment]
Parameter Golf competition deadline was April 30, 2026. Competition ended. Closing.
